library/lms.c

```diff
@@ -0,0 +1,777 @@
+
+    mbedtls_platform_zeroize( ctx, sizeof( mbedtls_lms_private_t ) );
+}
+
+
+int mbedtls_lms_generate_private_key( mbedtls_lms_private_t *ctx,
+                                      mbedtls_lms_algorithm_type_t type,
+                                      mbedtls_lmots_algorithm_type_t otstype,
+                                      int (*f_rng)(void *, unsigned char *, size_t),
+                                      void* p_rng, unsigned char *seed,
+                                      size_t seed_size )
+{
+    unsigned int idx = 0;
+    unsigned int free_idx = 0;
+    int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
+
+    if( type != MBEDTLS_LMS_SHA256_M32_H10 )
+    {
+        return( MBEDTLS_ERR_LMS_BAD_INPUT_DATA );
+    }
+
+    if( otstype != MBEDTLS_LMOTS_SHA256_N32_W8 )
+    {
+        return( MBEDTLS_ERR_LMS_BAD_INPUT_DATA );
+    }
+
+    if( ctx->have_private_key )
+    {
+        return( MBEDTLS_ERR_LMS_BAD_INPUT_DATA );
+    }
+
+    ctx->params.type = type;
+    ctx->params.otstype = otstype;
+
+    f_rng( p_rng,
+           ctx->params.I_key_identifier,
+           MBEDTLS_LMOTS_I_KEY_ID_LEN );
+
+    ctx->ots_private_keys = mbedtls_calloc( ( size_t )MERKLE_TREE_LEAF_NODE_AM(ctx->params.type),
+                                            sizeof( mbedtls_lmots_private_t ) );
+    if( ctx->ots_private_keys == NULL )
+    {
+        ret = MBEDTLS_ERR_LMS_ALLOC_FAILED;
+        goto exit;
+    }
+
+    ctx->ots_public_keys = mbedtls_calloc( ( size_t )MERKLE_TREE_LEAF_NODE_AM(ctx->params.type),
+                                           sizeof( mbedtls_lmots_public_t ) );
+    if( ctx->ots_public_keys == NULL )
+    {
+        ret = MBEDTLS_ERR_LMS_ALLOC_FAILED;
+        goto exit;
+    }
+
+    for( idx = 0; idx < MERKLE_TREE_LEAF_NODE_AM(ctx->params.type); idx++ )
+    {
+        mbedtls_lmots_init_private( &ctx->ots_private_keys[idx] );
+        mbedtls_lmots_init_public( &ctx->ots_public_keys[idx] );
+    }
+
+
+    for( idx = 0; idx < MERKLE_TREE_LEAF_NODE_AM(ctx->params.type); idx++ )
+    {
+        ret = mbedtls_lmots_generate_private_key( &ctx->ots_private_keys[idx],
+                                                  otstype,
+                                                  ctx->params.I_key_identifier,
+                                                  idx, seed, seed_size );
+        if( ret != 0 )
+            goto exit;
+
+        ret = mbedtls_lmots_calculate_public_key( &ctx->ots_public_keys[idx],
+                                                  &ctx->ots_private_keys[idx] );
+        if( ret != 0 )
+            goto exit;
+    }
+
+    ctx->q_next_usable_key = 0;
+    ctx->have_private_key = 1;
+
+exit:
+    if( ret != 0 )
+    {
+        for ( free_idx = 0; free_idx < idx; free_idx++ )
+        {
+            mbedtls_lmots_free_private( &ctx->ots_private_keys[free_idx] );
+            mbedtls_lmots_free_public( &ctx->ots_public_keys[free_idx] );
+        }
+
+        mbedtls_free( ctx->ots_private_keys );
```

Comment:
Directly calling mbedtls_lms_free_private causes issues when only one is allocated (since we can't distinguish between them. I've solved this by doing a manual deallocation in the second allocation failure handler, and changed where `have_private_key` is set so we properly free if the OTS key generation fails

Reply:
You can call the `xxx_free()` function safely if you maintain the invariant that `ctx->p` is either null or a valid pointer. This makes for a more robust architecture than having to keep track of when calling `xxx_free()` is safe and when it isn't. In any case, `xxx_free()` must be safe at any point except while an API function is working.